### PR TITLE
Checklists: Redux state, declarative structure

### PR DESCRIPTION
**Pull request by sirreal**
Implementing ideas from #26015
Supersedes #26030 
Closes #25958

This strikes a middle-ground of keeping array-of-objects that can be iterated over to produce a full checklist or a banner.

It also provides the option of building completely declarative checklists, which we do with the store checklist.

The props produced by `connect` are fewer and less likely to change.

Most importantly, we avoid merging _declarative, structural information_ with _state_.

This will be helpful for future progress, such as #25935 and #25954

Todo:

- ✅ Remove obsolete blocks
- ✅ Remove obsolete helpers (`mergeObjectIntoArrayById`)
- ✅ Fix classNames
- ✅ Update devdocs
- (postponed) Put [`launchTask`](https://github.com/Automattic/wp-calypso/pull/26216/files#diff-d07375849d476df77956914b2e2f5d1cR37) somewhere better (Redux actions or middleware)
     This could be handled in a follow-up, everything works in this PR, this is additional cleanup.

## Testing

Checklists may be used in 4 places. The functionality should remain unchanged for all of them:

Functionality:
- Task counts
- Task ordering (completed at the top)
- Checklist collapsing
- Clicking the event actions
- Reporting of event completion

### Checklists in use to test:

- [ ] Checklist banners (simple sites at `/stats/day/SITE`) (@taggon / @fditrapani)
- [ ] Jetpack checklist (Jetpack sites at `/plans/my-plan/SITE`) (@Automattic/luna)
- [x] Store checklist (Store installed, visit Store dashboard at `/store/SITE`) (@timmyc)
- [x] Devdocs https://calypso.live/devdocs/design/checklist?branch=update/checklist-separate-state-structure (@sirreal)
- [x] Devdocs playground (@sirreal): 
     Drop the following in https://calypso.live/devdocs/playground?branch=update/checklist-separate-state-structure
     ```jsx
     <Checklist inferCompletedCount>
     	<ChecklistTask
     		onClick={ () => console.log('Do this?') }
     		onDismiss={ () => console.log('Dismissing…') }
     		title="Reticulate splines"
     		buttonText="Reticulate!"
     		buttonPrimary
     		completedTitle="Splines are reticulated 👍"
     		description="Make sure that all the splines are reticulated."
     		duration="1 minute"
     		completed
     	/>
     	<ChecklistTask
     		onClick={ () => console.log('Do this?')  }
     		onDismiss={ () => console.log('Dismiss yak shaving.') }
     		title="Shave yaks!"
     		buttonText="Buzzzz"
     		buttonPrimary
     		completedTitle="Yaks shaved."
     		description="Make sure you shave the yaks so you can get on with your life."
     		duration="10,000 minutes"
     	/>
     </Checklist>
     ```

cc: @ockham @fditrapani

### CODE REVIEW

**Comment by fditrapani** [comment]
From a code perspective, this stuff is a little over my head. Let me know how I might be able to test it out when it's ready.

**Comment by sirreal** [comment]
@fditrapani This is ready for some testing. I've added some instructions about where the Checklists can be found for testing and what should be expected (same behavior as staging/production). I'd appreciate any testing feedback you can provide, thanks!

File: client/blocks/checklist/task.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	};
+
+	handleClick = () =>
+		/**
+		 * @TODO
+		 *
+		 * This should be a simpler dispatched action.
+		 * Look at moving logic into middleware.
+		 */
+		void launchTask( {
```
[the remark this answers is not in the document]
Comment (ockham) [reply client/blocks/checklist/task.js]: Yeah, I think ideally, I'd love to decouple the tour from the checklist even more -- i.e. really just pass in an `onClick` handler function, and make the `Task` component ignorant about GT specific stuff such as `tourSlug`, `tourUrl`, etc.

Follow-up material tho.

[the remark this answers is not in the document]
Comment (ockham) [reply client/blocks/checklist/task.js]: (Oh, which might mean we would possibly ditch the `Task` component (well, block) and try to encapsulate that bound `launchTask` function elsewhere...)

File: client/components/checklist/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	/>
+	<Task
+		onClick={ handleYakClick }
+		onDismiss={ handleYakDismiss }
+		title="Shave yaks!"
+		buttonText="Buzzzz"
+		buttonPrimary
+		completedTitle="Yaks shaved."
+		description="Make sure you shave the yaks so you can get on with your life."
+		duration="10,000 minutes"
```
Comment (ockham) [inline client/components/checklist/README.md]: :rofl:

File: client/blocks/checklist/task.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ockham) [reply client/blocks/checklist/task.js]: Okay, scrolling down to the `Task` _component_, I realize this is the major distinction between the component and the block. I think I'd still even go as far as to possibly remove the block altogether if we can find a different way of encapsulating the `launchTour` function...

**Comment by sirreal** [comment]
I've pinged a few folks for some manual testing as a checklist in the PR description.

If folks will contribute some manual testing to a checklist in their domain, @Automattic/luna or @ockham should be able to handle code review.

Thanks!

File: client/blocks/checklist/task.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sirreal) [reply client/blocks/checklist/task.js]: Yes, I agree with you here @ockham. I think this PR is a good balance of introducing the new, more declarative checklist _components_, and the `Task` _block_ may well be a temporary measure.

My inclination is that this PR is in a sweet spot right now, it introduces the components and maintains a connected block that makes the `state.checklist`-connected Tasks trivial to declare.

Finding the right place to handle tour start from the task (and whether that's implemented where/how it should be, or if it should be rethought entirely) can be part of a follow up.

What are your thoughts?

File: client/components/checklist/task.js [hunk 3]
```diff
@@ -95,19 +79,31 @@ export class ChecklistTask extends PureComponent {
...
 				</div>
-				<Focusable
+				<ToggleContainer
 					className="checklist__task-icon"
-					onClick={ this.handleToggle }
-					aria-pressed={ completed ? 'true' : 'false' }
+					onClick={ onDismiss }
+					aria-pressed={
+						/* eslint-disable-next-line no-nested-ternary */
+						onDismiss ? ( completed ? 'true' : 'false' ) : undefined
```
Comment (sirreal) [inline client/components/checklist/task.js]: Calling out for reviewers that ternaries are nested here in violation of eslint rules.

_My opinion_ is that this is quite a clear way to declare these props, but please raise any objections and I'll work out a change to drop the ternary nesting.

File: client/components/checklist/task.js [hunk 4]
```diff
@@ -95,19 +79,31 @@ export class ChecklistTask extends PureComponent {
...
+					}
 				>
 					<ScreenReaderText>
-						{ completed ? translate( 'Mark as uncompleted' ) : translate( 'Mark as completed' ) }
+						{ /* eslint-disable no-nested-ternary */
```
Comment (sirreal) [inline client/components/checklist/task.js]: Calling out for reviewers that ternaries are nested here in violation of eslint rules.

_My opinion_ is that this is quite a clear way to declare these props, but please raise any objections and I'll work out a change to drop the ternary nesting.

File: client/blocks/checklist/task.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ockham) [reply client/blocks/checklist/task.js]: Yep, sounds good! I think what seems the most arbitrary to me is that coupling of Guided Tours internals to the checklist. Agree that this PR is a good step towards decoupling that :+1:

File: client/components/checklist/task.js (hunk 3 above)
Comment (ockham) [reply client/components/checklist/task.js]: One thing you could consider is inlining the entire `<ToggleContainer />`, which might make things actually more readable:

```jsx
{ onDismiss
	? <Focusable
		onClick={ onDismiss }
		aria-pressed={ !! completed } // Do we really need the bool strings ('true'/'false') here?
	/>
	: <div />
}
```

**Comment by ockham** [comment]
I'm somewhat worried that having the `Checklist` component accept either a `Task` block or a `TaskComponent` might be confusing to consumers, and lead to misuse (and to some extent on the implementation side of `Checklist`, when it comes to extending and talking both cases into account).

In the long run, I think we should really get rid of that ambiguity. I really like the way the Woo checklist now looks -- this is still what I'd hope the dotcom and JP lists to eventually look like too.

I know the major obstacle is `ChecklistBanner` needing introspection. Is it possible to have e.g. `onboardingChecklist` export a prop that renders a list of `TaskComponent`s (plus some extra props, such as completed count, if needed)?

**Comment by ockham** [comment]
> Is it possible to have e.g. `onboardingChecklist` export a prop that renders a list of `TaskComponent`s (plus some extra props, such as completed count, if needed)?

Just tested that out (using #26030), that doesn't really work for our case:

Even if e.g. `OnboardingChecklist` renders an array of `Tasks`, and we use `const checklist = <OnboardingChecklist />` inside of `ChecklistBanner`: trying to `React.Children.map( checklist )` will still only iterate directly over the intermediate `OnboardingChecklist` (rather than its children).

**Comment by ockham** [comment]
Here's another idea tho to solve the `ChecklistBanner` issue:

Starting from #26030 way, we make `OnboardingChecklist` more generic by giving it another prop -- a component type, which it will use for individual checklist items. For the checklist banner scenario, we'll modify the `ChecklistBanner` component to be usable as such a checklist item (i.e. use props like `text`, `description`, `completed`). If `completed`, we have it render `null`.

```jsx
<OnboardingChecklist itemType={ ChecklistBanner }
```

Could that work? Is that reasonably ~elegant~ not too ugly?

Some more impl details:

* Probably easiest to start off #26030, since that already has checklists using `<Task />` components spelled out.
* No `ConnectedItem`s, `Task` blocks etc. Dumb `Item`s/`Task`s only. 
* WooCommerce checklist can stay largely as they are in both PRs now.
* Both JP and WP.com checklists: We manually set `completed` props at parent component level, just like in the Woo case.
* JP checklist can otherwise stay largely as-is in #26030 (since it's not used for a banner (yet)).

**Comment by sirreal** [comment]
> I'm somewhat worried that having the Checklist component accept either a `Task` block or a `TaskComponent` might be confusing to consumers, and lead to misuse (and to some extent on the implementation side of `Checklist`, when it comes to extending and talking both cases into account).
>
> In the long run, I think we should really get rid of that ambiguity. I really like the way the Woo checklist now looks -- this is still what I'd hope the dotcom and JP lists to eventually look like too.

I agree with all of this. The Woo checklist is what we want a checklist to look like 👍 

I really see the connected `Task` block as an intermediate step to removing any connected checklist blocks. That may fit nicely into this PR, I'll have a look…

> I know the major obstacle is `ChecklistBanner` needing introspection.

`ChecklistBanner` could use some work.

---

I want to reiterate, this PR doesn't complete the checklist work by any means. I think it strikes a good balance between being a bigger changeset than I'd normally like, but adding the necessary fundamental changes we need to continue work in this area. This opens the door for work rather than shutting it 🙂 

Fundamental bits:

- The Woo checklist it likely done and won't need to be touched.
- The (not connected) Checklist components have been reorganized under Components with their styles and are not likely to require major changes.

Needs work:

- ChecklistBanner
- ChecklistShow
- Try to drop connected Task

**Comment by sirreal** [comment]
rebased

**Comment by sirreal** [comment]
Please hold off on further reviews while I iterate on some ideas discussed with @ockham

File: client/components/checklist/README.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
+Checklist
+=======
+
+`Checklist` and `Task` are components used to render checklists.
+
+If your checklist tasks use checklist state from Redux, you can use `Task` from `blocks/checklist/tasks` to pull state by the `id` prop.
```
Comment (ockham) [inline client/components/checklist/README.md]: No moar `Task` block, is there?

**Comment by sirreal** [comment]
Closing in favor of #26618 and follow-up work.

**Pull request closed without merging**